### ISSUE TITLE
[Proposal] API redesign discussion

### ISSUE THREAD

**Issue by LonghronShen** [artiomchi/FlexLabs.Upsert#17]
API design proposal, as mentioned in PR #15 
This would make the API designs more semetically pretty.
```C#
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoUpdate((matched, newItem) => new Item()
    {
        Field = newItem.Field,
		Visit = matched.Visit + 1
    });
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoUpdateAsync((matched, newItem) => new Item()
    {
        Field = newItem.Field,
		Visit = matched.Visit + 1
    });
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoUpdate();
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoUpdateAsync();
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoNothing();
dbContext.Items.UpsertRange(newItems)
    .OnConflict(x => x.IdentitcalField)
    .DoNothingAsync();

// Infer the match fields to detected primary keys or alternative keys,
// this would make OnConflict funcional call optional
dbContext.Items.UpsertRange(newItems);
dbContext.Items.UpsertRangeAsync(newItems);
```
@artiomchi

**Comment by LonghronShen** [comment]
Duplicated with #16

**Comment by LonghronShen** [comment]
OK, good timing...

**Comment by artiomchi** [comment]
I definitely like the idea of a better semantic approach. Also, removing the Run would make sense.

The last option (`dbContext.Items.UpsertRange(newItems)`) wouldn't work though, since it returns a builder, and you'd need to actually trigger the call.

I thought that there might be better options for the syntax, since I don't want to align it to one database's syntax, excluding the others. Here are some of the other options we had in mind

* `OnConflict()` - has a somewhat negative connotation, and might not be the best option. While it aligns with the PostgreSQL syntax, that's not the case with any of the other databases.
* `OnMatch()` - is a little bit better, less aggressive, but also doesn't exactly explain what parameter it expects or what it does.
* `MatchOn()` - so far the most direct and obvious one. Database syntax agnostic, and very straightforward

I like the last one the most, but in that case, the second call doesn't quite make sense:
```c#
dbContext.Items.UpsertRange(newItems)
    .MatchOn(x => x.IdenticalField)
    .DoUpdate((matched, newItem) => new Item()
    {
        Field = newItem.Field,
		Visit = matched.Visit + 1
    });
```

Also, considering we can skip the match completely (to match on the PKs) and we want to add the no-update support (as long as the syntax allows it), it will start making even less sense (that's true no matter how we call the match method):
```c#
dbContext.Items.UpsertRange(newItems)
    .DoNothing();
```

**Comment by artiomchi** [comment]
After having some discussions about it, I actually came back close to what I started with :)

```c#
// It can be as simple as this, if you use the defaults (match on PK, update all)
dbContext.Items.Upsert(newItem).RunAsync();

// You can always use a custom match mapping
dbContext.Items.Upsert(newItem)
    .On(i => i.Date)
    .RunAsync();

// Pretty easy to specify which columns to update
dbContext.Items.Upsert(newItem)
    .On(i => i.Date)
    .WhenMatched(i => new Items { Visits = i.Visits + 1 })
    .RunAsync();

// And if you don't want to update existing entries
dbContext.Items.Upsert(newItem)
    .On(i => i.Date)
    .NoUpdate()
    .RunAsync();

// Or with the default match
dbContext.Items.Upsert(newItem).NoUpdate().RunAsync();
```

**Comment by LonghronShen** [comment]
Naming is not a problem, but the `Run` function is somewhat not EF style.
Another point, the `WhenMatched` function should support `Expression<Func<TEntity, TEntity, TEntity>>` to allow us to update according to the new item.

**Comment by artiomchi** [comment]
Well, the whole (and only) point of this discussion is coming up the naming and semantics :)

Regarding the change to the expression, I've actually opened up a new issue for this 👍

**Comment by artiomchi** [comment]
Regarding it not being EF style.. Most commands in EF don't actually run anything - they just add the entities to the change tracker. The changes are then committed when you call `SaveChanges()`. Compared to that, the `Upsert` command runs immediately on the server, outside of EF's change tracker, so in this context, having a `Run()` command actually makes sense, and is more explicit about what will happen

**Comment by LonghronShen** [comment]
At this point, we just support the upsert feature using ExecuteSqlCommand way, but in the future, is it possible to hook into the EF change tracker?

**Comment by artiomchi** [comment]
Well.. Yes and no..
Because of the way it runs, it's not really easy/possible to have EF trigger the commands, except for having to rewrite core parts of EF itself. So we'd have to run the upsert command as is, using `ExecureSqlCommand`. 

So at this point, unless we have any better ideas, I'll probably go ahead with the last option I described above.

Regarding the change tracker, we ran some tests, and we found that we can check EF's change tracker **after** the items have been inserted, and update the cached state of entities to the newly upserted values, to make sure the EF's cache is in sync with what we just did.
